src/clogger.cxx

```diff
@@ -309,3 +378,112 @@ log4cplus_logger_force_log_str(const log4cplus_char_t *name, loglevel_t ll,
 
     return retval;
 }
+
+static tstring const & customToStringMethod(LogLevel ll);
+static LogLevel customFromStringMethod(const tstring& nm);
+
+class CustomLogLevelmanager {
+    friend tstring const & customToStringMethod(LogLevel ll);
+    friend LogLevel customFromStringMethod(const tstring& nm);
+
+protected:
+    std::map<LogLevel,tstring> ll2nm;
+    std::map<tstring,LogLevel> nm2ll;
+    Hierarchy &h;
+
+public:
+    CustomLogLevelmanager()
+        : h(Logger::getDefaultHierarchy())
+    {
+        getLogLevelManager().pushToStringMethod(customToStringMethod);
+        getLogLevelManager().pushFromStringMethod(customFromStringMethod);
+    }
+
+    bool add(LogLevel ll, tstring const &nm)
+    {
+        // lock the DefaultHierarchy
+        HierarchyLocker theLock(h);
+        std::map<LogLevel,tstring>::iterator i = ll2nm.lower_bound(ll);
+        std::map<tstring,LogLevel>::iterator j = nm2ll.lower_bound(nm);
+        if( ( i != ll2nm.end() ) && ( i->first == ll ) && ( i->second != nm ) )
+            return false;
+        if( ( j != nm2ll.end() ) && ( j->first == nm ) && ( j->second != ll ) )
+            return false;
+
+        // there is no else after return
+        ll2nm.insert( i, std::make_pair(ll, nm) );
+        nm2ll.insert( j, std::make_pair(nm, ll) );
+        return true;
+    }
+
+    bool remove(LogLevel ll, tstring const &nm)
+    {
+        // lock the DefaultHierarchy
+        HierarchyLocker theLock(h);
+        std::map<LogLevel,tstring>::iterator i = ll2nm.find(ll);
+        std::map<tstring,LogLevel>::iterator j = nm2ll.find(nm);
+        if( ( i != ll2nm.end() ) && ( j != nm2ll.end() ) &&
+            ( i->first == j->second ) && ( i->second == j->first ) ) {
+            ll2nm.erase(i);
+            nm2ll.erase(j);
+
+            return true;
+        }
+
+        // there is no else after return
+        return false;
+    }
+};
+
+CustomLogLevelmanager customLogLevelmanager_;
+static log4cplus::tstring const empty_str;
```

Comment:
We already have one such [`empty_str`](https://github.com/log4cplus/log4cplus/blob/2.0.x/include/log4cplus/internal/internal.h#L70) in `internal` namespace.

Reply:
Static variables shouldn't affect any other compile unit. If it does, we might have a serious problem.

Reply:
I mean, you can use the other `empty_str` instance instead of introducing a new one.